core/services/keeper/upkeep_executer.go

```diff
@@ -239,51 +239,69 @@ func (executer *UpkeepExecuter) execute(upkeep UpkeepRegistration, headNumber in
+	pipeline.PromPipelineTaskExecutionTime.WithLabelValues(fmt.Sprintf("%d", ex.job.ID), ex.job.Name.String, "", job.Keeper.String()).Set(float64(elapsed))
 	var status string
 	if runErrors.HasError() || err != nil {
 		status = "error"
-		pipeline.PromPipelineRunErrors.WithLabelValues(fmt.Sprintf("%d", executer.job.ID), executer.job.Name.String).Inc()
+		pipeline.PromPipelineRunErrors.WithLabelValues(fmt.Sprintf("%d", ex.job.ID), ex.job.Name.String).Inc()
 	} else {
 		status = "completed"
 	}
-	pipeline.PromPipelineRunTotalTimeToCompletion.WithLabelValues(fmt.Sprintf("%d", executer.job.ID), executer.job.Name.String).Set(float64(elapsed))
-	pipeline.PromPipelineTasksTotalFinished.WithLabelValues(fmt.Sprintf("%d", executer.job.ID), executer.job.Name.String, "", job.Keeper.String(), status).Inc()
+	pipeline.PromPipelineRunTotalTimeToCompletion.WithLabelValues(fmt.Sprintf("%d", ex.job.ID), ex.job.Name.String).Set(float64(elapsed))
+	pipeline.PromPipelineTasksTotalFinished.WithLabelValues(fmt.Sprintf("%d", ex.job.ID), ex.job.Name.String, "", job.Keeper.String(), status).Inc()
 }
 
-func (executer *UpkeepExecuter) constructCheckUpkeepCallMsg(upkeep UpkeepRegistration) (ethereum.CallMsg, error) {
+func (ex *UpkeepExecuter) constructCheckUpkeepCallMsg(upkeep UpkeepRegistration) (ethereum.CallMsg, error) {
 	checkPayload, err := RegistryABI.Pack(
 		checkUpkeep,
 		big.NewInt(int64(upkeep.UpkeepID)),
 		upkeep.Registry.FromAddress.Address(),
 	)
 	if err != nil {
-		return ethereum.CallMsg{}, err
+		return ethereum.CallMsg{}, errors.Wrap(err, "failed to pack payload of check upkeep")
 	}
 
 	to := upkeep.Registry.ContractAddress.Address()
-	gasLimit := executer.config.KeeperRegistryCheckGasOverhead() + uint64(upkeep.Registry.CheckGas) +
-		executer.config.KeeperRegistryPerformGasOverhead() + upkeep.ExecuteGas
-	msg := ethereum.CallMsg{
+	gasLimit := ex.config.KeeperRegistryCheckGasOverhead() + uint64(upkeep.Registry.CheckGas) +
+		ex.config.KeeperRegistryPerformGasOverhead() + upkeep.ExecuteGas
+
+	return ethereum.CallMsg{
 		From: utils.ZeroAddress,
 		To:   &to,
 		Gas:  gasLimit,
 		Data: checkPayload,
+	}, nil
+}
+
+// logRevertReason logs the given error with a log level depends on this given error context
+// Default log level is error. Mapping between a reason and its log level:
+//  - "upkeep not needed": Debug
+func (ex *UpkeepExecuter) logRevertReason(logger *logger.Logger, err error) {
+	revertReason, err2 := eth.ExtractRevertReasonFromRPCError(err)
+	if err2 != nil {
+		logger.WithError(err).Errorf("UpkeepExecuter: checkUpkeep call failed with unknown reason, err2: %v", err2)
+		return
 	}
 
-	return msg, nil
+	logger = logger.With("revertReason", revertReason)
+	switch revertReason {
+	case "upkeep not needed":
+		logger.Debug("UpkeepExecuter: checkUpkeep call failed with known reason")
+	default:
+		logger.Error("UpkeepExecuter: checkUpkeep call failed with known reason")
```

Comment:
I'm worried this will create too many false-positive error level logs. We would essentially need a case for every revert reason in the [registry contract](https://github.com/smartcontractkit/keeper/blob/master/contracts/KeeperRegistry.sol). And some of them would have to be matched by regex, since they wrap customer contract revert reasons. I think adding that is a bit overkill.

Reply:
I think "upkeep not needed" is not actually an error so I would not log that with "error" level. The same for "keepers must take turns". These are just messages and they are not really important for us so I would apply "debug" log level for these messages. And if the incoming message is not one of those two, I would log it as an error. Does it make sense?

Reply:
I've just moved those two messages to constants. If more such messages will be coming, I would think about another solution.